doc/api.rst

```diff
@@ -577,6 +577,143 @@ The ``monadUserState-bytestring`` wrapper is the same as the ``monadUserState``
+They can also be somewhat faster to process.
+Note that using these wrappers adds a dependency on the ``Text`` modules, which live in the ``text`` package.
+
+Note also that by default tokens share space with the input ``Text`` which has the advantage that it does not need to make a copy but it also prevents the input from being garbage collected.
+It may make sense in some applications to use ``Text``'s ``copy`` function to unshare tokens that will be kept for a long time, to allow the original input to be collected.
+
+The "strict-text" wrapper
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``strict-text`` wrapper is the same as the ``basic`` wrapper but with strict ``Text`` instead of ``String``:
+
+.. code-block:: haskell
+
+   import           Data.Text (Text)
+   import qualified Data.Text as Text
+
+   type AlexInput =
+     ( Char      -- previous char
+     , [Byte]    -- rest of the bytes for the current char
+     , Text      -- rest of the input Text
+     )
+
+   alexGetByte       :: AlexInput -> Maybe (Char, AlexInput)
+
+   alexInputPrevChar :: AlexInput -> Char
+
+   -- alexScanTokens :: ByteString -> [token]
+
+All of the actions in your lexical specification should have type:
+
+.. code-block:: haskell
+
+   { ... } :: Text -> token
+
+for some type ``token``.
+
+The "posn-strict-text" wrapper
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``posn-strict-text`` wrapper is the same as the ``posn`` wrapper but with strict ``Text`` instead of ``String``:
+
+.. code-block:: haskell
+
+   import           Data.Text (Text)
+   import qualified Data.Text as Text
+
+   type AlexInput =
+     ( AlexPosn      -- current position,
+     , Char          -- previous char
+     , [Byte]        -- rest of the bytes for the current char
+     , Text          -- current input Text
+     )
+
+   -- alexScanTokens :: Text -> [token]
+
+All of the actions in your lexical specification should have type:
+
+.. code-block:: haskell
+
+   { ... } :: AlexPosn -> Text -> token
+
+for some type ``token``.
+
+The "monad-strict-text" wrapper
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``monad-strict-text`` wrapper is the same as the ``monad`` wrapper but with strict ``Text`` instead of ``String``:
+
+.. code-block:: haskell
+
+   import           Data.Text (Text)
+   import qualified Data.Text as Text
+
+   data AlexState = AlexState
+     { alex_pos   :: !AlexPosn  -- position at current input location
+     , alex_inp   :: Text       -- the current input
+     , alex_chr   :: !Char      -- the character before the input
+     , alex_bytes :: [Byte]     -- rest of the bytes for the current char
+     , alex_scd   :: !Int       -- the current startcode
+     }
+
+   newtype Alex a = Alex { unAlex :: AlexState
+                                  -> Either String (AlexState, a) }
+
+   instance Functor     Alex where ...
+   instance Applicative Alex where ...
+   instance Monad       Alex where ...
```

Comment:
Do we need these `instance` stubs?  
(In practice, one would use some automatic deriving tactics like `GeneralizedNewtypeDeriving` or `DerivingVia`...)

Reply:
I think the piece of code is supposed to show users what the types are that are being generated, copied from above.

Reply:
Ok, fair enough.  (Didn't see them for the `bytestring` wrapper, but nothing wrong having them here.)